File: src/main/java/io/github/binout/soccer/domain/season/SeasonPlanning.java
package io.github.binout.soccer.domain.season;

import io.github.binout.soccer.domain.date.*;
import io.github.binout.soccer.domain.event.FriendlyMatchPlanned;
import io.github.binout.soccer.domain.event.LeagueMatchPlanned;
import io.github.binout.soccer.domain.player.Player;
import io.github.binout.soccer.domain.player.PlayerRepository;
import io.github.binout.soccer.domain.season.match.FriendlyMatch;
import io.github.binout.soccer.domain.season.match.LeagueMatch;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Component;

import java.time.LocalDate;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Component
public class SeasonPlanning {

    private final PlayerRepository playerRepository;
    private final FriendlyMatchDateRepository friendlyMatchDateRepository;
    private final LeagueMatchDateRepository leagueMatchDateRepository;
    private final MatchPlanning matchPlanning;
    private final ApplicationEventPublisher publisher;

    @Autowired
    public SeasonPlanning(PlayerRepository playerRepository,
                          FriendlyMatchDateRepository friendlyMatchDateRepository,
                          LeagueMatchDateRepository leagueMatchDateRepository,
                          MatchPlanning matchPlanning,
                          ApplicationEventPublisher publisher) {
        this.playerRepository = playerRepository;
        this.friendlyMatchDateRepository = friendlyMatchDateRepository;
        this.leagueMatchDateRepository = leagueMatchDateRepository;
        this.matchPlanning = matchPlanning;
        this.publisher = publisher;
    }

    public LeagueMatch planLeagueMatch(Season season, LeagueMatchDate date) {
        TreeMap<Integer, List<Player>> treeMap = computeGamesPlayed(date, playerRepository.all().filter(Player::isPlayerLeague), MatchPlanning.leagueCounter(season));
        LeagueMatch leagueMatch = season.addLeagueMatch(date, extractPlayers(treeMap, LeagueMatch.MAX_PLAYERS, true));
        publisher.publishEvent(new LeagueMatchPlanned(leagueMatch, matchPlanning.getSubstitutes(season, leagueMatch)));
        return leagueMatch;
    }

    public FriendlyMatch planFriendlyMatch(Season season, FriendlyMatchDate date) {
        TreeMap<Integer, List<Player>> treeMap = computeGamesPlayed(date, playerRepository.all(), MatchPlanning.globalCounter(season));
        FriendlyMatch friendlyMatch = season.addFriendlyMatch(date, extractPlayers(treeMap, FriendlyMatch.MAX_PLAYERS, false));
        publisher.publishEvent(new FriendlyMatchPlanned(friendlyMatch, matchPlanning.getSubstitutes(season, friendlyMatch)));
        return friendlyMatch;
    }

    private TreeMap<Integer, List<Player>> computeGamesPlayed(MatchDate date, Stream<Player> players, Function<Player, Integer> counter) {
        Map<Integer, List<Player>> playersByGamesPlayed = players.filter(date::isPresent).collect(Collectors.groupingBy(counter));
        return new TreeMap<>(playersByGamesPlayed);
    }

    private Set<Player> extractPlayers(TreeMap<Integer, List<Player>> treeMap, int maxPlayers, boolean goalPriority) {
        Set<Player> players = new HashSet<>();
        if (goalPriority) {
            treeMap.values().stream().flatMap(Collection::stream).filter(Player::isGoalkeeper).forEach(players::add);
        }
        Iterator<Map.Entry<Integer, List<Player>>> iterator = treeMap.entrySet().iterator();
        while (iterator.hasNext() && teamIsNotFull(players, maxPlayers)) {
            Map.Entry<Integer, List<Player>> entry = iterator.next();
            List<Player> playersForCurrentCount = entry.getValue();
            Collections.shuffle(playersForCurrentCount); // randomize list
            Iterator<Player> playerIterator = playersForCurrentCount.iterator();
            while (playerIterator.hasNext() && teamIsNotFull(players, maxPlayers)) {
                players.add(playerIterator.next());
            }
        }
        return players;
    }

    private boolean teamIsNotFull(Set<Player> players, int maxPlayers) {
        return players.size() < maxPlayers;
    }


    public List<FriendlyMatchDate> friendlyMatchDatesToPlan(Season season) {
        Set<LocalDate> dates = season.friendlyMatches().map(FriendlyMatch::date).collect(Collectors.toSet());
        return matchDatesToPlan(dates, friendlyMatchDateRepository.all());
    }

    public List<LeagueMatchDate> leagueMatchDatesToPlan(Season season) {
        Set<LocalDate> dates = season.leagueMatches().map(LeagueMatch::date).collect(Collectors.toSet());
        return matchDatesToPlan(dates, leagueMatchDateRepository.all());
    }

    private <T extends MatchDate> List<T> matchDatesToPlan(Set<LocalDate> dates, Stream<T> allDates) {
        LocalDate now = LocalDate.now();
        return allDates
                .filter(d -> d.date().isAfter(now) || d.date().isEqual(now))
                .filter(d -> !dates.contains(d.date()))
                .filter(MatchDate::canBePlanned)
                .collect(Collectors.toList());
    }


}

File: src/main/kotlin/io/github/binout/soccer/domain/season/SeasonPlanning.kt
/*
 * Copyright 2016 Benoît Prioux
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.binout.soccer.domain.season

import io.github.binout.soccer.domain.date.*
import io.github.binout.soccer.domain.FriendlyMatchPlanned
import io.github.binout.soccer.domain.LeagueMatchPlanned
import io.github.binout.soccer.domain.player.Player
import io.github.binout.soccer.domain.player.PlayerRepository

import java.time.LocalDate
import java.util.*
import javax.enterprise.context.ApplicationScoped
import javax.enterprise.event.Event
import javax.inject.Inject

@ApplicationScoped
class SeasonPlanning() {

    private var seasonRepository: SeasonRepository? = null
    private var playerRepository: PlayerRepository? = null
    private var friendlyMatchDateRepository: FriendlyMatchDateRepository? = null
    private var leagueMatchDateRepository: LeagueMatchDateRepository? = null
    private var matchPlanning: MatchPlanning? = null

    @Inject
    var friendlyMatchPlannedPublisher: Event<FriendlyMatchPlanned>? = null
    @Inject
    var leagueMatchPlannedPublisher: Event<LeagueMatchPlanned>? = null

    @Inject
    constructor(seasonRepository: SeasonRepository,
                playerRepository: PlayerRepository,
                friendlyMatchDateRepository: FriendlyMatchDateRepository,
                leagueMatchDateRepository: LeagueMatchDateRepository,
                matchPlanning: MatchPlanning) : this() {
        this.seasonRepository = seasonRepository
        this.playerRepository = playerRepository
        this.friendlyMatchDateRepository = friendlyMatchDateRepository
        this.leagueMatchDateRepository = leagueMatchDateRepository
        this.matchPlanning = matchPlanning
    }

    fun planLeagueMatch(season: Season, date: LeagueMatchDate): LeagueMatch {
        val treeMap = computeGamesPlayed(date, playerRepository!!.all().filter { it.isPlayerLeague }, MatchPlanning.leagueCounter(season))
        val leagueMatch = season.addLeagueMatch(date, extractPlayers(treeMap, LeagueMatch.MAX_PLAYERS, true))
        seasonRepository!!.replace(season)
        leagueMatchPlannedPublisher?.fire(LeagueMatchPlanned(leagueMatch, matchPlanning!!.getSubstitutes(season, leagueMatch)))
        return leagueMatch
    }

    fun planFriendlyMatch(season: Season, date: FriendlyMatchDate): FriendlyMatch {
        val treeMap = computeGamesPlayed(date, playerRepository!!.all(), MatchPlanning.globalCounter(season))
        val friendlyMatch = season.addFriendlyMatch(date, extractPlayers(treeMap, FriendlyMatch.MAX_PLAYERS, false))
        seasonRepository!!.replace(season)
        friendlyMatchPlannedPublisher?.fire(FriendlyMatchPlanned(friendlyMatch, matchPlanning!!.getSubstitutes(season, friendlyMatch)))
        return friendlyMatch
    }

    private fun computeGamesPlayed(date: MatchDate, players: List<Player>, counter: (Player) -> Int): TreeMap<Int, List<Player>> =
            TreeMap(players.filter { date.isPresent(it) }.groupBy(counter))

    private fun extractPlayers(treeMap: TreeMap<Int, List<Player>>, maxPlayers: Int, goalPriority: Boolean): Set<Player> {
        val players = HashSet<Player>()
        if (goalPriority) {
            treeMap.values.flatten().filter { it.isGoalkeeper }.forEach { players.add(it) }
        }
        val iterator = treeMap.entries.iterator()
        while (iterator.hasNext() && teamIsNotFull(players, maxPlayers)) {
            val entry = iterator.next()
            val playersForCurrentCount = entry.value.shuffled()
            val playerIterator = playersForCurrentCount.iterator()
            while (playerIterator.hasNext() && teamIsNotFull(players, maxPlayers)) {
                players.add(playerIterator.next())
            }
        }
        return players
    }

    private fun teamIsNotFull(players: Set<Player>, maxPlayers: Int): Boolean = players.size < maxPlayers

    fun friendlyMatchDatesToPlan(season: Season): List<FriendlyMatchDate> {
        val dates = season.friendlyMatches().map { it.date }.toSet()
        return matchDatesToPlan(dates, friendlyMatchDateRepository!!.all())
    }

    fun leagueMatchDatesToPlan(season: Season): List<LeagueMatchDate> {
        val dates = season.leagueMatches().map { it.date }.toSet()
        return matchDatesToPlan(dates, leagueMatchDateRepository!!.all())
    }

    private fun <T : MatchDate> matchDatesToPlan(dates: Set<LocalDate>, allDates: List<T>): List<T> {
        val now = LocalDate.now()
        return allDates
                .filter { d -> d.date.isAfter(now) || d.date.isEqual(now) }
                .filter { d -> !dates.contains(d.date) }
                .filter { it.canBePlanned() }
    }


}
